### ISSUE TITLE
Database Cleanup and Sequelize Model <> DB Schema Synchronization

### ISSUE THREAD

**Issue by timolegros** [hicommonwealth/commonwealth#3349]
EDIT: See https://github.com/hicommonwealth/commonwealth/issues/3349#issuecomment-1560896760

## ~~Description~~
~~Tables such as the `Comments` table have duplicate and incorrectly named indexes. The most common naming issue is~~ ~~indexes named `offchain_[table_name]_...`. These indexes were not renamed during the transition away from the legacy~~ ~~`Offchain_...` table naming scheme.~~

### ~~Sample of Changes (old - see sub tickets for exact changes)~~
~~**Addresses**~~
- ~~`models/Address.ts` indicates index on `name` -> missing in production~~

~~**Comments**~~
- ~~`offchain_comments_address_id` -> removed since `comments_address_id` already exists.~~
- ~~`offchain_commnets_chain_object_id` -> removed since `comments_chain_object_id` already exists.~~
- ~~`comments_id` -> removed since there exists an indexed primary key on the same `id` column.~~
- ~~`offchain_comments_chain_created_at` -> renamed to `comments_chain_created_at`.~~
- ~~`offchain_comments_chain_updated_at` -> renamed to `comments_chain_updated_at`.~~
- ~~`offchain_comments_id` -> removed since primary key exists.~~
- ~~`offchain_comments_root_id` -> renamed to `comments_root_id` -> can this index be removed?~~
- ~~`OffchainComments_search` -> renamed to `comments_search` -> can this index be removed?~~
- ~~The `models/comment.ts` model indicates that there should be an index on the `canvas_hash` column but none exists -> should one be created?~~
- ~~The `models/comment.ts` model indicates that there should be an index on the `thread_id` column but none exists -> should one be created?~~
- ~~The `models/comment.ts` model indicates that there should be an index on the `(chain, thread_id)` but none exists -> should one be created?~~

~~**Reactions**~~
- ~~Remove `offchain_...` prefix from all indexes.~~
- ~~Remove `offchain_reactions_id` -> primary key exists~~
- ~~`canvas_hash` index is missing in production -> should it be created?~~

## Acceptance Criteria
- No duplicate indexes
- Proper nomenclature

**Comment by timolegros** [comment]
@NakulManchanda This ticket will open up after #3317 is merged

**Comment by timolegros** [comment]
Root id refactor removed the `root_id` in the `Comment.ts` Sequelize model but not from the database schema so I am removing the `offchain_comments_root_id` index along with the column.

**Comment by NakulManchanda** [comment]
Hi @timolegros 
yes, i looked at `comments` table too
- its ok to group all index for one table in one migration script/transaction
    - we can lock table once and make all the changes
- verify migration time run on firck/frack
- not sure what could be rollback strategy, hopefully we don't need one

**Comment by NakulManchanda** [comment]
- Renaming is fine not necessary per say, but if we are doing its fine
-  "offchain_comments_root_id" btree (root_id) - remove and add index for `thread_id` that be most important change here
- feel free to sync model definition with database definition - for canvas changes




<img width="1229" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/2e0aa183-9662-49bf-a7b0-f572169cfac7">

**Comment by NakulManchanda** [comment]
**Addresses** - not a big change feel free to include this - just verify index exists in prod
- Migration script `CREATE INDEX IF NOT EXISTS` won't harm 
  - so that all environments can benefit - if you like, and model definition remains true for all
  
**Reactions**
 - i see we are pulling reactions by address_id, chain, proposal_id, thread_id, comment_id basically everything
 - feel free to **add index for proposal_id as well**

**Comment by NakulManchanda** [comment]
@timolegros 
One PR would be fine for all three - address, comments and reactions.
Feel free to split into different migration script - so not to lock all three tables at once.

**Comment by timolegros** [comment]
I strikethrough the issue above as the description is outdated. This issue is now a bucket ticket for database cleanup and sequelize model <> DB schema synchronization. Here is the basic gist of the sub-PRs:

1. Remove `offchain` prefixes from all indexes and keys.
  - While not necessary, we have **not** used tables with `Offchain` prefixes for months so for the sake of clarity and consistency and in the spirit of cleaning up the codebase this will be cleaned up.
2. Fix constraint inconsistencies.
  - We have many cases where a constraint such as `NOT NULL` is set in the Sequelize model but not in the DB schema or vice versa.
3. Add/Remove Indexes
  - We have numerous duplicate and/or unnecessary indexes in the DB that can and **should** be removed because they slow down write operations.
  - We also have numerous indexes that are defined in Sequelize models but were never created via a migration. If the index defined in the model is beneficial (performance-wise) then it needs to be created.
4. Remove old/unused columns and tables
  - `OldChatMessages` is an old table that we have no use for anymore
  - Several columns in a variety of tables seem to be unused and we may be able to drop them.

**Comment by NakulManchanda** [comment]
Just 
- use `Alter` instead of `Drop` wherever possible
- add **if exists** wherever possible and needed
- group by related tables in one PR eg adresss, comments, reactions
- measure performance of each migration on frick/frack

Thanks for update above!

**Comment by NakulManchanda** [comment]
@timolegros - for this you can create other issue too, rather than clubbing it here!

for **Chat** we can do separate PR for sure
- i see references like this models.Chain.hasMany(models.ChatChannel);
- remove websocket code too if not needed

do it as last PR, after all the changes above!

**Comment by timolegros** [comment]
> @timolegros - for this you can **create other issue too, rather than clubbing it here**!
> 
> for **Chat** we can do separate PR for sure
> 
> * i see references like this models.Chain.hasMany(models.ChatChannel);
> * remove websocket code too if not needed anymore (websocket, if not used remove some load from redis/server connections too i believe)
> 
> It has **DB cleanup** but it has **application cleanup too** i believe if not already done! do it as last PR, after all the changes above, though feel free to lay foundations, by using some columns etc if you like!

- Yea waiting on a response from Jake/Forest on whether I should create separate issues and move the into 'in-progress'.
- `OldChatMessages` is an old table unrelated to the newer `ChatMessages` table.
- `Sequelize Model` in the title covers app side changes (no other changes will be made to the app/client)
- WebSocket code is used for notifications so it will not be removed at this time.

**Comment by NakulManchanda** [comment]
cool, thanks for another update!
- sequelize model update is part of DB cleanup only, sounds good for chat!
- so if only these index / constraint changes and we already know which one to target, one issue would be fine
ofcourse, defer to you guys! Thanks again!
Keep going with PR's issues can be created retrospectively too!

**Comment by NakulManchanda** [comment]
for indexes, i like targeting queries approach
- please make note of any improvement you will see when deploy on frick etc!
- before & after time of queries

**Comment by NakulManchanda** [comment]
Sorry, @timolegros you are best person to take it through!
I did one separate PR for **"Comments" table** change - https://github.com/hicommonwealth/commonwealth/pull/4083

Would be happy to review when back, feel free to reopen - 
https://github.com/hicommonwealth/commonwealth/pull/3729 - way too big
https://github.com/hicommonwealth/commonwealth/pull/3975 - this one should be quick when back!

Really, like the research and cleanup effort - just kind of focused on performance related indexes - also in middle of changing schema for Notifications/Subscriptions etc

**Comment by timolegros** [comment]
This issue will get split into multiple issues associated with a single milestone.
